pkg/webhook/elasticquota/quota_topology.go

```diff
@@ -0,0 +1,664 @@
+	if !exist || len(children) == 0 {
+		// this quotaGroup doesn't have childQuota
+		return true
+	}
+
+	childMinSum, err := qt.getChildMinQuotaSum(quotaInfo.Name)
+	if err != nil {
+		klog.Errorf("checkMinQuotaSum failed:%v", err)
+	}
+
+	if notLarger, _ := quotav1.LessThanOrEqual(childMinSum, quotaInfo.CalculateInfo.OriginalMin); !notLarger {
+		klog.Errorf("checkMinQuotaSum allChild SumMinQuota > MinQuota %v", quotaInfo.Name)
+		return false
+	}
+
+	return true
+}
+
+func (qt *quotaTopology) getChildMinQuotaSum(parentName string) (allChildQuotaSum v1.ResourceList, err error) {
+	allChildQuotaSum = v1.ResourceList{}
+	if parentName == extension.RootQuotaName {
+		return allChildQuotaSum, nil
+	}
+
+	childQuotaList, exist := qt.quotaHierarchyInfo[parentName]
+	if !exist {
+		return nil, fmt.Errorf("not found child quota list:%v", parentName)
+	}
+
+	for childName := range childQuotaList {
+		if quotaInfo, exist := qt.quotaInfoMap[childName]; exist {
+			allChildQuotaSum = quotav1.Add(allChildQuotaSum, quotaInfo.CalculateInfo.OriginalMin)
+		} else {
+			err = fmt.Errorf("BUG quotaInfoMap and quotaTree information out of sync, losed :%v", childName)
+			klog.Error(err)
+			return nil, err
+		}
+	}
+
+	return allChildQuotaSum, nil
+}
+
+func (qt *quotaTopology) checkMaxQuota(quotaInfo *core.QuotaInfo) bool {
+	if quotaInfo.ParentName != extension.RootQuotaName {
+		parentInfo, exist := qt.quotaInfoMap[quotaInfo.ParentName]
+		if !exist {
+			klog.Errorf("checkQuotaMax %v not found parentInfo:%v", quotaInfo.Name, quotaInfo.ParentName)
+			return false
+		}
+		for key, val := range parentInfo.CalculateInfo.Max {
+			childMax := quotaInfo.CalculateInfo.Max[key]
+			if childMax.Cmp(val) == 1 {
+				klog.Errorf("checkQuotaMax maxQuota(%v) > parentMaxQuota %v(%v)", quotaInfo.CalculateInfo.Max, quotaInfo.ParentName, parentInfo.CalculateInfo.Max)
+				return false
+			}
+		}
+	}
+
+	children, exist := qt.quotaHierarchyInfo[quotaInfo.Name]
+	if !exist || len(children) == 0 {
+		return true
+	}
+
+	// maxQuota should be >= each child's maxQuota
+	for childName := range children {
+		if childQuotaInfo, exist := qt.quotaInfoMap[childName]; exist {
+			for key, val := range quotaInfo.CalculateInfo.Max {
+				childMax := childQuotaInfo.CalculateInfo.Max[key]
+				if childMax.Cmp(val) == 1 {
+					klog.Errorf("checkQuotaMax child %v(%v) maxQuota > self maxQuota(%v)", childName, childQuotaInfo.CalculateInfo.Max, quotaInfo.CalculateInfo.Max)
+					return false
+				}
+			}
+		} else {
+			klog.Errorf("BUG quotaInfoMap and quotaTree information out of sync, losed :%v", childName)
+			return false
+		}
+	}
+	return true
+}
+
+func (qt *quotaTopology) checkQuotaHierarchyHasChange(oldQuotaInfo, newQuotaInfo *core.QuotaInfo) error {
+	// parentQuota -> childQuota, check whether the quotaInfo has child or not
+	if oldQuotaInfo.IsParent == true && newQuotaInfo.IsParent == false {
+		childQuotaSet, exist := qt.quotaHierarchyInfo[newQuotaInfo.Name]
```

Comment:
1. if from child->parent, need check child quota don't have pods.
2.if pod submit to parent, reject
3.in check function,  don't change quotaHierarchyInfo, check is a read action